pkg/etcdcli/etcdcli.go

```diff
@@ -69,86 +64,67 @@ func NewEtcdClient(kubeInformers v1helpers.KubeInformersForNamespaces, networkIn
-		if net.ParseIP(bootstrapIP).To4() == nil {
-			bootstrapIP = "[" + bootstrapIP + "]"
+		var etcdEndpoints []string
+		nodes, err := g.nodeLister.List(labels.Set{"node-role.kubernetes.io/master": ""}.AsSelector())
+		for _, node := range nodes {
+			internalIP, err := dnshelpers.GetEscapedPreferredInternalIPAddressForNodeName(network, node)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get internal IP for node: %w", err)
+			}
+			etcdEndpoints = append(etcdEndpoints, fmt.Sprintf("https://%s:2379", internalIP))
 		}
-		etcdEndpoints = append(etcdEndpoints, fmt.Sprintf("https://%s:2379", bootstrapIP))
-	}
-
-	g.clientLock.Lock()
-	defer g.clientLock.Unlock()
-	if reflect.DeepEqual(g.lastClientConfigKey, etcdEndpoints) && g.cachedClient != nil {
-		ctx, cancel := context.WithCancel(context.Background())
-		defer cancel()
-		// Test cached client connection by doing a request.
-		_, err = g.cachedClient.MemberList(ctx)
-		if err == nil {
-			return g.cachedClient, nil
+
+		configmap, err := g.configmapsLister.ConfigMaps(operatorclient.TargetNamespace).Get("etcd-endpoints")
+		if err != nil {
+			return nil, fmt.Errorf("failed to list endpoints from %s/%s: %w",
+				operatorclient.TargetNamespace, "etcd-endpoints", err)
 		}
+		if bootstrapIP, ok := configmap.Annotations[BootstrapIPAnnotationKey]; ok && bootstrapIP != "" {
+			// escape if IPv6
+			if net.ParseIP(bootstrapIP).To4() == nil {
+				bootstrapIP = "[" + bootstrapIP + "]"
+			}
+			etcdEndpoints = append(etcdEndpoints, fmt.Sprintf("https://%s:2379", bootstrapIP))
+		}
+		return etcdEndpoints, nil
 	}
 
-	c, err := getEtcdClientWithClientOpts(etcdEndpoints, false)
-	if err != nil {
-		return nil, fmt.Errorf("failed to get etcd client: %w", err)
-	}
-	if g.cachedClient != nil {
-		if err := g.cachedClient.Close(); err != nil {
-			utilruntime.HandleError(fmt.Errorf("failed to close cached client: %w", err))
+	newFunc := func() (*clientv3.Client, error) {
+		endpoints, err := endpointFunc()
+		if err != nil {
+			return nil, fmt.Errorf("error retrieving endpoints for new cached client: %w", err)
 		}
+		return newEtcdClientWithClientOpts(endpoints, true)
 	}
-	g.cachedClient = c
-	g.lastClientConfigKey = etcdEndpoints
 
-	return g.cachedClient, nil
+	g.clientPool = NewDefaultEtcdClientPool(newFunc, endpointFunc)
+	return g
 }
 
-func getEtcdClientWithClientOpts(endpoints []string, skipConnectionTest bool, opts ...ClientOption) (*clientv3.Client, error) {
+// newEtcdClientWithClientOpts allows customization of the etcd client using ClientOptions. All clients must be manually
+// closed by the caller with Close().
+func newEtcdClientWithClientOpts(endpoints []string, skipConnectionTest bool, opts ...ClientOption) (*clientv3.Client, error) {
 	grpclog.SetLoggerV2(grpclog.NewLoggerV2(ioutil.Discard, ioutil.Discard, os.Stderr))
 	clientOpts, err := newClientOpts(opts...)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error during clientOpts: %w", err)
 	}
```

Comment:
Good addition of clearer error messages throughout. Much appreciated!

Reply:
thank @Elbehery!